### PR TITLE
refactor(c): Refactor indent queries

### LINKED ISSUE #4079: Indent with curly braces on new line behaves weirdly
### Describe the bug

When placing an opening curly braces on newline right after for/while, vim/nvim will (try to) align it with the indentation of the line above, but not for treesitter indent

### To Reproduce

1. Open nvim with a C parser, opening the file below
```c
#include <stdio.h>

int main()
{
  while(1)
       // put the cursor here
}
```
2. Enter insert mode, insert `{`

### Expected behavior

The code should be like below
```c
#include <stdio.h>

int main()
{
  while(1)
  {
}
```

### Actual behavior

The curly brace is placed at where the cursor is, not trying to put itself at the position that vim/nvim will place it
```c
#include <stdio.h>

int main()
{
  while(1)
       {
}
```

### Output of `:checkhealth nvim-treesitter`

```markdown
nvim-treesitter: require("nvim-treesitter.health").check()

Installation ~
- OK `tree-sitter` found 0.20.7 (b268e412ad4848380166af153300464e5a1cf83f) (parser generator, only needed for :TSInstallFromGrammar)
- OK `node` found v16.15.0 (only needed for :TSInstallFromGrammar)
- OK `git` executable found.
- OK `cc` executable found. Selected from { vim.NIL, "cc", "gcc", "clang", "cl", "zig" }
  Version: cc (GCC) 12.2.0
- OK Neovim was compiled with tree-sitter runtime ABI version 14 (required >=13). Parsers must be compatible with runtime ABI.

OS Info:
{
  machine = "x86_64",
  release = "6.1.1-zen1-1-zen",
  sysname = "Linux",
  version = "#1 ZEN SMP PREEMPT_DYNAMIC Wed, 21 Dec 2022 22:27:59 +0000"
} ~

Parser/Features         H L F I J
  - todotxt             ✓ . . . .
  - lua                 ✓ ✓ ✓ ✓ ✓
  - java                ✓ ✓ . ✓ ✓
  - php                 ✓ ✓ ✓ ✓ ✓
  - query               ✓ ✓ ✓ ✓ ✓
  - mermaid             ✓ . . . .
  - markdown            ✓ . ✓ . ✓
  - c                   ✓ ✓ ✓ ✓ ✓
  - css                 ✓ . ✓ ✓ ✓
  - tsx                 ✓ ✓ ✓ ✓ ✓
  - gitignore           ✓ . . . .
  - sxhkdrc             ✓ . ✓ . ✓
  - yaml                ✓ ✓ ✓ ✓ ✓
  - pug                 ✓ . . . ✓
  - svelte              ✓ . ✓ ✓ ✓
  - rasi                ✓ ✓ ✓ ✓ .
  - jsonc               ✓ ✓ ✓ ✓ ✓
  - scss                ✓ . ✓ ✓ .
  - scheme              ✓ . ✓ . ✓
  - javascript          ✓ ✓ ✓ ✓ ✓
  - rust                ✓ ✓ ✓ ✓ ✓
  - html                ✓ ✓ ✓ ✓ ✓
  - r                   ✓ ✓ . ✓ ✓
  - typescript          ✓ ✓ ✓ ✓ ✓
  - awk                 ✓ . . . ✓
  - jsdoc               ✓ . . . .
  - diff                ✓ . . . .
  - comment             ✓ . . . .
  - toml                ✓ ✓ ✓ ✓ ✓
  - gitcommit           ✓ . . . ✓
  - gitattributes       ✓ . . . ✓
  - python              ✓ ✓ ✓ ✓ ✓
  - cpp                 ✓ ✓ ✓ ✓ ✓
  - bash                ✓ ✓ ✓ . ✓
  - zig                 ✓ . ✓ ✓ ✓
  - regex               ✓ . . . .
  - git_rebase          ✓ . . . ✓
  - markdown_inline     ✓ . . . ✓
  - help                ✓ . . . ✓
  - vue                 ✓ . ✓ ✓ ✓
  - json                ✓ ✓ ✓ ✓ .
  - vim                 ✓ ✓ ✓ . ✓

  Legend: H[ighlight], L[ocals], F[olds], I[ndents], In[j]ections
         +) multiple parsers found, only one will be used
         x) errors found in the query, try to run :TSUpdate {lang} ~
```

### Output of `nvim --version`

```text
NVIM v0.9.0-dev-594+gc590641fe
Build type: Release
LuaJIT 2.1.0-beta3
Compiled by root@spacegliders

Features: +acl +iconv +tui
See ":help feature-compile"

   system vimrc file: "$VIM/sysinit.vim"
  fall-back for $VIM: "/usr/local/share/nvim"

Run :checkhealth for more info
```

### Additional context

This may not only affect C but probably many other filetypes as well

### LINKED ISSUE #4117: Else Statement in c++ is not indented
### Describe the bug

When you complete writing the if else statement in c++, it removes the indentation from the else statement, I know indentation is still experimental, I just wanted to get bring this to attention

### To Reproduce

type
if(){
}
else{
}
then 

### Expected behavior

else statement should not be unindented

### Output of `:checkhealth nvim-treesitter`

```markdown
nvim-treesitter: require("nvim-treesitter.health").check()
========================================================================
## Installation
  - OK: `tree-sitter` found 0.20.7 (b268e412ad4848380166af153300464e5a1cf83f) (parser generator, only needed for :TSInstallFromGrammar)
  - OK: `node` found v18.12.1 (only needed for :TSInstallFromGrammar)
  - OK: `git` executable found.
  - OK: `cc` executable found. Selected from { vim.NIL, "cc", "gcc", "clang", "cl", "zig" }
    Version: cc (Ubuntu 9.4.0-1ubuntu1~20.04.1) 9.4.0
  - OK: Neovim was compiled with tree-sitter runtime ABI version 14 (required >=13). Parsers must be compatible with runtime ABI.

## OS Info:
{
  machine = "x86_64",
  release = "5.15.79.1-microsoft-standard-WSL2",
  sysname = "Linux",
  version = "#1 SMP Wed Nov 23 01:01:46 UTC 2022"
}

## Parser/Features         H L F I J
  - vim                 ✓ ✓ ✓ . ✓
  - lua                 ✓ ✓ ✓ ✓ ✓
  - cpp                 ✓ ✓ ✓ ✓ ✓
  - c                   ✓ ✓ ✓ ✓ ✓
  - python              ✓ ✓ ✓ ✓ ✓
  - help                ✓ . . . ✓

  Legend: H[ighlight], L[ocals], F[olds], I[ndents], In[j]ections
         +) multiple parsers found, only one will be used
         x) errors found in the query, try to run :TSUpdate {lang}
```

### Output of `nvim --version`

```text
NVIM v0.8.2-dev-da68023
Build type: Release
LuaJIT 2.1.0-beta3
```

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by lucario387**
Resolves #4079
Resolves #4117
Behavior changes:
- Before:
  - An if statement without code blocks, following by another if statement/ code blocks will be indented twice
  - If the chain continues, then it is indented multiple times (equal to the amount of bare if statements that appears in the chain of if-else above it)
- Now:
  - A bare if statement, followed by a bare if statement, then the second bare if statement will not be indented any extra
I have yet to cover for the situations below
- else immediately followed by a for/while/switch statement (indent twice)
  - This case I don't know how to handle, as there may be two styles for this (for/while in the same/different line to else)

<details>
<summary>Supported if styles</summary>   

```   
if 
  ret
else if
  ret
else
  ret

if
{
  ret
}
else if
{
  ret
}

if {
  ret
} else if {
  ret
} else {
  ret
}

if {
  ret
}
else if {
  ret
}
else {
  ret
}
```
</details>

### CODE REVIEW

File: queries/c/indents.scm [hunk 1]
```diff
@@ -1,28 +1,51 @@
...
   (field_declaration_list)
   (case_statement)
   (enumerator_list)
-  (struct_specifier)
   (compound_literal_expression)
   (initializer_list)
-  (while_statement)
-  (for_statement)
-  (switch_statement)
+  (init_declarator)
```
Comment (lucario387) [inline queries/c/indents.scm:8]: This will be the most controversial part anyways, so cc-ing @theHamsta 
Even with the new queries, all the tests still passed, so does this mean weak tests or over-coverage of the queries before?

File: queries/c/indents.scm [hunk 2]
```diff
@@ -1,28 +1,74 @@
...
+(
+  (while_statement
+    body: (_) @_body
+  ) @indent
+  (#not-has-type? @_body compound_statement)
+)
+
 
-(compound_statement "}" @indent_end)
+(compound_statement "}" @indent_end)
```
Comment (theHamsta) [inline queries/c/indents.scm:69]: white space at EOL

File: queries/c/indents.scm (hunk 1 above)
Comment (theHamsta) [reply queries/c/indents.scm:8]: there were some problem with `compound_statement` only the last time I tried, but I would just judge based on a net-positive on the tests. When we discover new bugs with the new queries we can revert or find a better solution.

File: queries/c/indents.scm [hunk 3]
```diff
@@ -1,28 +1,74 @@
 [
-  (compound_statement)
-  (preproc_arg)
```
Comment (theHamsta) [inline queries/c/indents.scm:3]: is `preproc_arg` related to this change? If I remember correctly it is for indenting second lines of macros. Not sure if this is tested in the tests.

Comment (lucario387) [reply queries/c/indents.scm:3]: Somewhere below of this file, someone marks (preproc_arg) as @ignore, so I removed it. 
Can add it back if you think is needed

**Comment by lucario387** [comment]
Okay, I managed to get all the if cases working as expected
Should be good for a second look now @theHamsta

File: queries/c/indents.scm (hunk 3 above)
Comment (lucario387) [reply queries/c/indents.scm:3]: https://github.com/nvim-treesitter/nvim-treesitter/blob/master/queries/c/indents.scm#L42
It is marked as `@ignore` right afterwards, which will cancel the `@indent`, so I removed it

File: queries/c/indents.scm [hunk 4]
```diff
@@ -1,28 +1,81 @@
...
+        (if_statement) 
+        (if_statement consequence: (compound_statement) @dedent)
+      ] @dedent
+  )@indent
+  (#not-has-type? @_consequence compound_statement)
+)
+
+(
+  ERROR
+    "for" "(" @indent ";" ";" ")" @indent_end)
```
Comment (lucario387) [inline queries/c/indents.scm:56]: This is to maintain similar behavior for `for` compared to current master branch behavior

File: tests/indent/c/if_else.c [hunk 5]
```diff
@@ -0,0 +1,30 @@
+int foo(int x){
+    if (x > 10)
```
Comment (theHamsta) [inline tests/indent/c/if_else.c:2]: do we have enough tests for if-compound statement and stronger nestings?

**Review by theHamsta — APPROVED** [review]

File: tests/indent/c/if_else.c (hunk 5 above)
Comment (lucario387) [reply tests/indent/c/if_else.c:2]: I don't think so, but it is unnecessary because even before this PR, in the case of if-compound statements, the if statements were never indented to begin with, it was handled by the compound statement

Comment (lucario387) [reply tests/indent/c/if_else.c:2]: > do we have enough tests for if-compound_statement+different combinations and stronger nestings (ifs in ifs)

Tested locally and did not seem to have issues, I will push some extra tests to check coverage for it

Comment (lucario387) [reply tests/indent/c/if_else.c:2]: Ooops, miss one case

Comment (lucario387) [reply tests/indent/c/if_else.c:2]: Tests added

**Comment by lucario387** [comment]
Nested ifs are chores, finding all the edge cases tomorrow

**Comment by lucario387** [comment]
Have yet to cover for
else with for/while/switch statements

**Comment by lucario387** [comment]
@jdrouhard @theHamsta please have a look!
I think i have covered for most of the if-else cases possible

**Comment by theHamsta** [comment]
Amazing! Thank you a lot!
